File: 5 semestre/Programacao para Dispositivos Moveis/UAS/UA1/java/src/empresa/Funcionario .java
package empresa;

import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.UUID;

public abstract class Funcionario implements Bonus {
    protected String id;
    protected String nome;
    protected String cpf;
    protected double salarioBase;
    protected Departamento departamento;
    protected LocalDate dataAdmissao;
    protected boolean ativo;

    public Funcionario(String nome, String cpf, double salarioBase, Departamento departamento, LocalDate dataAdmissao) {
        this.id = UUID.randomUUID().toString().substring(0, 8);
        this.nome = nome;
        this.cpf = cpf;
        this.salarioBase = salarioBase;
        this.departamento = departamento;
        this.dataAdmissao = dataAdmissao;
        this.ativo = true;
    }

    public abstract double calcularSalarioTotal();

    public void desativar() {
        this.ativo = false;
    }

    public void ativar() {
        this.ativo = true;
    }

    public String getId() {
        return id;
    }

    public String getNome() {
        return nome;
    }

    public String getCpf() {
        return cpf;
    }

    public double getSalarioBase() {
        return salarioBase;
    }

    public Departamento getDepartamento() {
        return departamento;
    }

    public LocalDate getDataAdmissao() {
        return dataAdmissao;
    }

    public boolean isAtivo() {
        return ativo;
    }

    public void aumentarSalario(double percentual) {
        if (percentual > 0) {
            this.salarioBase += this.salarioBase * (percentual / 100);
        }
    }

    @Override
    public String toString() {
        DateTimeFormatter fmt = DateTimeFormatter.ofPattern("dd/MM/yyyy");
        return String.format("%s | %s | %s | R$ %.2f | %s | Adm: %s | Ativo: %s",
                id, nome, cpf, salarioBase, departamento, dataAdmissao.format(fmt), ativo);
    }
}

File: 5 semestre/Programacao para Dispositivos Moveis/UAS/UA1/dart/empresa/funcionario.dart
abstract class Funcionario implements Bonus {
  final String          id;
  final String          nome;
  final String          cpf;
  final Departamento    departamento;
  final DateTime        dataAdmissao;
  double                salarioBase;
  bool                  ativo;

  Funcionario({
    required this.nome,
    required this.cpf,
    required this.salarioBase,
    required this.departamento,
    required this.dataAdmissao,
  })  : id = _gerarId(),
        ativo = true;

  static String _gerarId() {
    const chars = 'abcdefghijklmnopqrstuvwxyz0123456789';
    Random rnd = Random();
    return String.fromCharCodes(
      Iterable.generate(8, (_) => chars.codeUnitAt(rnd.nextInt(chars.length))),
    );
  }

  double calcularSalarioTotal();

  void desativar()  => ativo = false;
  void ativar()     => ativo = true;

  void aumentarSalario(double percentual) {
    if (percentual > 0) {
      salarioBase += salarioBase * (percentual / 100);
    }
  }

  @override
  String toString() {
    final fmt = "dd/MM/yyyy";
    final dataFormatada = "${dataAdmissao.day.toString().padLeft(2, '0')}/"
        "${dataAdmissao.month.toString().padLeft(2, '0')}/"
        "${dataAdmissao.year}";
    return "$id | $nome | $cpf | R\$ ${salarioBase.toStringAsFixed(2)} | $departamento | Adm: $dataFormatada | Ativo: $ativo";
  }
}